bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/client/Client.java

```diff
@@ -0,0 +1,434 @@
+        return locations;
+    }
+
+    /**
+     * Parse FMI multipointcoverage formatted xml response
+     *
+     * @param response
+     * @return
+     * @throws FMIResponseException
+     * @throws SAXException
+     * @throws IOException
+     * @throws XPathExpressionException
+     */
+    private FMIResponse parseMultiPointCoverageXml(String response)
+            throws FMIResponseException, SAXException, IOException, XPathExpressionException {
+        Document document = documentBuilder.parse(new InputSource(new StringReader(response)));
+
+        XPath xPath = XPathFactory.newInstance().newXPath();
+        xPath.setNamespaceContext(NAMESPACE_CONTEXT);
+
+        boolean isExceptionReport = ((Node) xPath.compile("/ows:ExceptionReport").evaluate(document,
+                XPathConstants.NODE)) != null;
+        if (isExceptionReport) {
+            Node exceptionCode = (Node) xPath.compile("/ows:ExceptionReport/ows:Exception/@exceptionCode")
+                    .evaluate(document, XPathConstants.NODE);
+            String[] exceptionText = queryNodeValues(xPath.compile("//ows:ExceptionText/text()"), document);
+            throw new FMIExceptionReportException(exceptionCode.getNodeValue(), exceptionText);
+        }
+
+        Builder builder = new FMIResponse.Builder();
+
+        String[] parameters = queryNodeValues(xPath.compile("//swe:field/@name"), document);
+        /**
+         * Observations have FMISID (FMI Station ID?), with forecasts we use lat & lon
+         */
+        String[] ids = queryNodeValues(xPath.compile(
+                "//target:Location/gml:identifier[@codeSpace='http://xml.fmi.fi/namespace/stationcode/fmisid']/text()"),
+                document);
+
+        String[] names = queryNodeValues(xPath.compile(
+                "//target:Location/gml:name[@codeSpace='http://xml.fmi.fi/namespace/locationcode/name']/text()"),
+                document);
+        String[] representativePointRefs = queryNodeValues(
+                xPath.compile("//target:Location/target:representativePoint/@xlink:href"), document);
+
+        if ((ids.length > 0 && ids.length != names.length) || names.length != representativePointRefs.length) {
+            throw new FMIResponseException(String.format(
+                    "Could not all properties of locations: ids: %d, names: %d, representativePointRefs: %d",
+                    ids.length, names.length, representativePointRefs.length));
+        }
+
+        Location[] locations = new Location[representativePointRefs.length];
+        for (int i = 0; i < locations.length; i++) {
+            BigDecimal[] latlon = findLatLon(xPath, i, document, representativePointRefs[i]);
+            String id = ids.length == 0 ? String.format("%s,%s", latlon[0].toPlainString(), latlon[1].toPlainString())
+                    : ids[i];
+            locations[i] = new Location(names[i], id, latlon[0], latlon[1]);
+        }
+
+        logger.trace("names ({}): {}", names.length, names);
```

Comment:
Are these trace messages useful for any user or could you use a debugger instead?

Reply:
I would like to leave these to troubleshoot user issues (results depend on time, cannot necessarily be easily reproduced by a developer afterwards)